### PR TITLE
TEST: added mpi alltoall[v] test

### PR DESCRIPTION
**Pull request by karasevb**

### CODE REVIEW

**Comment by karasevb** [comment]
@vspetrov @Sergei-Lebedev please review

**Comment by karasevb** [comment]
In progress. Not ready for review yet.

**Comment by karasevb** [comment]
Ready for review.

File: test/mpi/test_mpi.h [hunk 1]
```diff
@@ -187,6 +232,61 @@ class TestBcast : public TestCase {
...
+    size_t rncounts;
+    int *scounts;
+    int *sdispls;
+    int *rcounts;
+    int *rdispls;
+    ucc_test_vsize_flag_t count_bits;
+    ucc_test_vsize_flag_t displ_bits;
+
+    template<typename T>
+    static void* _mpi_counts_to_ucc(int *mpi_counts, size_t _ncount)
```
Comment (vspetrov) [inline test/mpi/test_mpi.h:257]: lets move implementations into test_alltoallv.cc ?

File: test/mpi/test_mpi.h [hunk 2]
```diff
@@ -187,6 +232,61 @@ class TestBcast : public TestCase {
...
+public:
+    TestAlltoallv(size_t _msgsize, ucc_test_mpi_inplace_t inplace,
+                  ucc_datatype_t _dt, ucc_memory_type_t _mt,
+                  ucc_test_team_t &_team,
+                  ucc_test_vsize_flag_t count_bits,
+                  ucc_test_vsize_flag_t displ_bits);
+    ucc_status_t check();
+    std::string str();
+    ~TestAlltoallv()
+    {
```
Comment (vspetrov) [inline test/mpi/test_mpi.h:274]: sam here

File: test/mpi/test_alltoall.cc [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        UCC_CHECK(ucc_mc_alloc(&check_sbuf, _msgsize * nprocs, UCC_MEMORY_TYPE_HOST));
+        init_buffer(check_sbuf, count * nprocs, dt, _mt, rank);
+    } else {
+        args.mask = UCC_COLL_ARGS_FIELD_FLAGS;
+        args.flags = UCC_COLL_ARGS_FLAG_IN_PLACE;
+        init_buffer(rbuf, count * nprocs, dt, _mt, rank);
+        init_buffer(check_rbuf, count * nprocs, dt, UCC_MEMORY_TYPE_HOST, rank);
+    }
+
+    args.coll_type            = UCC_COLL_TYPE_ALLTOALL;
```
[the remark this answers is not in the document]
Comment (karasevb) [reply test/mpi/test_alltoall.cc:42]: If some test was skipped, the coll type should be specified before skip() call, it required in std::string str();

File: test/mpi/main.cc [hunk 4]
```diff
@@ -3,19 +3,29 @@
 #include "test_mpi.h"
 #include <chrono>
 
+int test_rand_seed = -1;
+size_t test_max_size = 0;
```
Comment (vspetrov) [inline test/mpi/main.cc:7]: maybe make "max_size" part of the test class.
Also, lets make non 0 default, say 8 Mb. Then, e.g. running on the test on 1K ranks, will mean we can still run allgather up to 8K msg size per proc.

File: test/mpi/buffer.cc [hunk 5]
```diff
@@ -65,6 +72,11 @@ void init_buffer(void *buf, size_t count, ucc_datatype_t dt,
         MPI_Abort(MPI_COMM_WORLD, -1);
         break;
     }
+    if (UCC_MEMORY_TYPE_HOST != mt) {
+        UCC_CHECK(ucc_mc_memcpy(_buf, buf, count * ucc_dt_size(dt),
```
Comment (Sergei-Lebedev) [inline test/mpi/buffer.cc:76]: maybe use UCC_ALLOC_COPY_BUF as in compare_buffers

File: test/mpi/main.cc [hunk 6]
```diff
@@ -236,9 +268,41 @@ static void process_root(const char *arg)
...
+    std::cout << "\n===== UCC MPI TEST INFO =======\n"
+              << "   seed        : " << std::to_string(test_rand_seed) << "\n"
+              <<   "===============================\n"
+              << std::endl;
+}
+
 void ProcessArgs(int argc, char** argv)
 {
-    const char* const short_opts = "c:t:m:d:o:M:I:r:h";
+    const char* const short_opts = "c:t:m:d:o:M:I:r:s:C:D:Z:h";
```
Comment (Sergei-Lebedev) [inline test/mpi/main.cc:305]: need to check somewhere that requested memory type is available in ucc, perhaps we can do it in mtype_str_to_type

File: test/mpi/test_allgather.cc [hunk 7]
```diff
@@ -19,21 +19,30 @@ TestAllgather::TestAllgather(size_t _msgsize, ucc_test_mpi_inplace_t _inplace,
...
+    }
+
     UCC_CHECK(ucc_mc_alloc(&rbuf, _msgsize*size, _mt));
-    UCC_CHECK(ucc_mc_alloc(&check_buf, _msgsize*size, _mt));
+    UCC_CHECK(ucc_mc_alloc(&check_rbuf, _msgsize*size, UCC_MEMORY_TYPE_HOST));
     if (TEST_NO_INPLACE == inplace) {
         UCC_CHECK(ucc_mc_alloc(&sbuf, _msgsize, _mt));
         init_buffer(sbuf, count, TEST_DT, _mt, rank);
+        UCC_CHECK(ucc_mc_alloc(&check_sbuf, _msgsize, UCC_MEMORY_TYPE_HOST));
+        init_buffer(check_sbuf, count, TEST_DT, UCC_MEMORY_TYPE_HOST, rank);
```
Comment (Sergei-Lebedev) [inline test/mpi/test_allgather.cc:36]: why not to use UCC_ALLOC_COPY_BUF ?  we just need to copy from check_sbuf to sbuf
same question for other tests

File: test/mpi/test_allgather.cc [hunk 8]
```diff
@@ -19,21 +19,30 @@ TestAllgather::TestAllgather(size_t _msgsize, ucc_test_mpi_inplace_t _inplace,
...
     } else {
         args.mask = UCC_COLL_ARGS_FIELD_FLAGS;
         args.flags = UCC_COLL_ARGS_FLAG_IN_PLACE;
         init_buffer((void*)((ptrdiff_t)rbuf + rank*count*dt_size),
                     count, TEST_DT, _mt, rank);
-        init_buffer((void*)((ptrdiff_t)check_buf + rank*count*dt_size),
-                    count, TEST_DT, _mt, rank);
+        init_buffer((void*)((ptrdiff_t)check_rbuf + rank*count*dt_size),
```
Comment (Sergei-Lebedev) [inline test/mpi/test_allgather.cc:42]: UCC_ALLOC_COPY_BUF ?

File: test/mpi/test_alltoallv.cc [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+    args.src.info_v.buffer = sbuf;
+    args.src.info_v.datatype = _dt;
+    args.src.info_v.mem_type = _mt;
+
+    args.dst.info_v.buffer = rbuf;
+    args.dst.info_v.datatype = _dt;
+    args.dst.info_v.mem_type = _mt;
+
+    if (TEST_FLAG_VSIZE_64BIT == count_bits) {
+        args.src.info_v.counts =
```
Comment (Sergei-Lebedev) [inline test/mpi/test_alltoallv.cc:104]: _mpi_count_to_ucc  allocates memory,  where do you call free() for it?

**Review by Sergei-Lebedev — CHANGES_REQUESTED** [review]

File: test/mpi/buffer.cc (hunk 5 above)
Comment (karasevb) [reply test/mpi/buffer.cc:76]: UCC_ALLOC_COPY_BUF not quite suitable here, `init_buffer()` not required to allocate buf. Please fix me if I'm wrong.

File: test/mpi/test_alltoallv.cc (hunk 9 above)
Comment (karasevb) [reply test/mpi/test_alltoallv.cc:118]: fixed

File: test/mpi/test_allgather.cc (hunk 7 above)
Comment (karasevb) [reply test/mpi/test_allgather.cc]: fixed

File: test/mpi/main.cc (hunk 6 above)
Comment (karasevb) [reply test/mpi/main.cc:315]: fixed

File: test/mpi/main.cc (hunk 4 above)
Comment (karasevb) [reply test/mpi/main.cc]: fixed

File: test/mpi/test_mpi.h (hunk 2 above)
Comment (karasevb) [reply test/mpi/test_mpi.h]: fixed

File: test/mpi/test_mpi.h (hunk 1 above)
Comment (karasevb) [reply test/mpi/test_mpi.h]: fixed

File: test/mpi/test_case.cc [hunk 10]
```diff
@@ -73,38 +88,76 @@ ucc_status_t TestCase::exec()
...
     return status;
 }
 
+test_skip_cause_t TestCase::skip(int skip_cond, test_skip_cause_t cause,
+                                 MPI_Comm comm)
+{
+    int rank;
+    test_skip_cause_t skip = skip_cond ? cause : TEST_SKIP_NONE;
+    MPI_Comm_rank(comm, &rank);
+    MPI_Reduce((void*)&skip, (void*)&test_skip, 1,
```
[the remark this answers is not in the document]
Comment (karasevb) [reply test/mpi/test_case.cc]: fixed

File: test/mpi/test_alltoallv.cc [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+                (ucc_count_t*)_mpi_counts_to_ucc<uint64_t>(scounts, nprocs);
+        args.dst.info_v.counts =
+                (ucc_count_t*)_mpi_counts_to_ucc<uint64_t>(rcounts, nprocs);
+    } else {
+        args.src.info_v.counts =
+                (ucc_count_t*)_mpi_counts_to_ucc<uint32_t>(scounts, nprocs);
+        args.dst.info_v.counts =
+                (ucc_count_t*)_mpi_counts_to_ucc<uint32_t>(rcounts, nprocs);
+    }
+    if (TEST_FLAG_VSIZE_64BIT & displ_bits) {
```
[the remark this answers is not in the document]
Comment (karasevb) [reply test/mpi/test_alltoallv.cc]: fixed

File: test/mpi/test_allgatherv.cc [hunk 12]
```diff
@@ -16,15 +16,24 @@ TestAllgatherv::TestAllgatherv(size_t _msgsize, ucc_test_mpi_inplace_t _inplace,
     size_t dt_size = ucc_dt_size(TEST_DT);
     size_t count = _msgsize/dt_size;
     int rank, size;
+    counts = NULL;
+    displacements = NULL;
     MPI_Comm_rank(team.comm, &rank);
     MPI_Comm_size(team.comm, &size);
+    args.coll_type                = UCC_COLL_TYPE_ALLGATHERV;
+
+    if (skip(test_max_size &&
```
[the remark this answers is not in the document]
Comment (karasevb) [reply test/mpi/test_allgatherv.cc]: fixed

File: test/mpi/test_alltoall.cc [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+
+ucc_status_t TestAlltoall::check()
+{
+    size_t count = args.src.info.count;
+    MPI_Alltoall(inplace ? MPI_IN_PLACE : check_sbuf, count, ucc_dt_to_mpi(dt),
+                 check_rbuf, count, ucc_dt_to_mpi(dt), team.comm);
+    return compare_buffers(rbuf, check_rbuf, count, dt, mem_type);
+}
+
+std::string TestAlltoall::str()
```
[the remark this answers is not in the document]
Comment (karasevb) [reply test/mpi/test_alltoall.cc]: fix

File: test/mpi/test_allgather.cc [hunk 14]
```diff
@@ -19,21 +19,30 @@ TestAllgather::TestAllgather(size_t _msgsize, ucc_test_mpi_inplace_t _inplace,
     MPI_Comm_rank(team.comm, &rank);
     MPI_Comm_size(team.comm, &size);
 
+    args.coll_type         = UCC_COLL_TYPE_ALLGATHER;
+
+    if (skip(test_max_size &&
```
[the remark this answers is not in the document]
Comment (karasevb) [reply test/mpi/test_allgather.cc]: fixed

File: test/mpi/buffer.cc [hunk 15]
```diff
@@ -84,23 +96,37 @@ ucc_status_t compare_buffers_fp(T *b1, T *b2, size_t count) {
     return UCC_OK;
 }
 
-ucc_status_t compare_buffers(void *rst, void *expected, size_t count,
+ucc_status_t compare_buffers(void *_rst, void *expected, size_t count,
                              ucc_datatype_t dt, ucc_memory_type_t mt)
 {
     ucc_status_t status = UCC_ERR_NO_MESSAGE;
+    void *rst = NULL;
+    int rst_alloced = 0;
```
[the remark this answers is not in the document]
Comment (karasevb) [reply test/mpi/buffer.cc]: fix

File: test/mpi/test_alltoall.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (karasevb) [reply test/mpi/test_alltoall.cc]: fixed

File: test/mpi/test_allgather.cc (hunk 8 above)
Comment (karasevb) [reply test/mpi/test_allgather.cc:43]: not quite suitable for `check_rbuf`, in both cases it is pre-allocated.

File: test/mpi/main.cc [hunk 16]
```diff
@@ -98,15 +117,25 @@ static ucc_coll_type_t coll_str_to_type(std::string coll)
...
+        mem_type = UCC_MEMORY_TYPE_CUDA;
     } else {
         std::cerr << "incorrect memory type: " << mtype << std::endl;
         PrintHelp();
     }
-    abort();
+    if (UCC_MEMORY_TYPE_HOST != mem_type && UCC_OK != ucc_mc_available(mem_type)) {
+        std::cerr << "requested memory type '"
+                  << ucc_memory_type_names[mem_type]
+                  << "' is not support "
```
Comment (vspetrov) [inline test/mpi/main.cc:133]: not supportED, also there are extra " ' " single quotes.

File: test/mpi/test_bcast.cc [hunk 17]
```diff
@@ -10,23 +10,31 @@
 #define TEST_DT UCC_DT_UINT32
 
 TestBcast::TestBcast(size_t _msgsize, ucc_memory_type_t _mt,
-                     int _root, ucc_test_team_t &_team) :
-    TestCase(_team, _mt, _msgsize)
+                     int _root, ucc_test_team_t &_team, size_t _max_size) :
+    TestCase(_team, _mt, _msgsize, (ucc_test_mpi_inplace_t)0, _max_size)
```
Comment (vspetrov) [inline test/mpi/test_bcast.cc:14]: TEST_NO_INPLACE?

**Review by vspetrov — APPROVED** [review]

**Comment by vspetrov** [comment]
@Sergei-Lebedev @karasevb guys, what is left there ? lets complete this one?

File: test/mpi/main.cc (hunk 16 above)
Comment (karasevb) [reply test/mpi/main.cc]: fixed

File: test/mpi/test_bcast.cc (hunk 17 above)
Comment (karasevb) [reply test/mpi/test_bcast.cc]: fixed

**Review by Sergei-Lebedev — APPROVED** [review]

**Comment by karasevb** [comment]
@vspetrov ready to go

**Pull request merged**